File: Sources/Api/CustomObjectAPI.m
//
//  CustomObjectAPI.m
//  LoginRadiusSDK
//
//  Created by LoginRadius on 13/12/17.
//

#import "CustomObjectAPI.h"

@implementation CustomObjectAPI

+ (instancetype)customInstance {
    static dispatch_once_t onceToken;
    static CustomObjectAPI *instance;
    
    dispatch_once(&onceToken, ^{
        instance = [[CustomObjectAPI alloc] init];
    });
    
    return instance;
}

- (void)createCustomObjectWithObjectName:(NSString*)objectname
                       accessToken:(NSString*)accessToken
                              payload:(NSDictionary*)payload
                 completionHandler:(LRAPIResponseHandler)completion{
    
    [[LoginRadiusREST apiInstance] sendPOST:@"identity/v2/auth/customobject"
                                queryParams:@{
                                              @"apikey": [LoginRadiusSDK apiKey],
                                              @"access_token": accessToken,
                                              @"objectname": objectname
                                              }
                                       body:payload
                          completionHandler:completion];
}

- (void)getCustomObjectWithObjectName:(NSString*)objectname
                    accessToken:(NSString*)accessToken
              completionHandler:(LRAPIResponseHandler)completion{
    
    [[LoginRadiusREST apiInstance] sendGET:@"identity/v2/auth/customobject"
                               queryParams:@{
                                             @"apikey": [LoginRadiusSDK apiKey],
                                             @"access_token": accessToken,
                                             @"objectname": objectname
                                             }
                         completionHandler:completion];
}

- (void)getCustomObjectWithObjectRecordId:(NSString*)objectRecordId
                    accessToken:(NSString*)accessToken
                 objectname:(NSString*)objectname
              completionHandler:(LRAPIResponseHandler)completion{
    
    NSString* url = [NSString stringWithFormat:@"identity/v2/auth/customobject/%@", objectRecordId];
    [[LoginRadiusREST apiInstance] sendGET:url
                               queryParams:@{
                                             @"apikey": [LoginRadiusSDK apiKey],
                                             @"access_token": accessToken,
                                             @"objectname": objectname
                                             }
                         completionHandler:completion];
}

- (void)updateCustomObjectWithObjectName:(NSString*)objectname
                    accessToken:(NSString*)accessToken
                 objectRecordId:(NSString*)objectRecordId
                 updatetype:(NSString*)updatetype
                payload:(NSDictionary*)payload
              completionHandler:(LRAPIResponseHandler)completion{
   
    NSString* url = [NSString stringWithFormat:@"identity/v2/auth/customobject/%@", objectRecordId];
    [[LoginRadiusREST apiInstance] sendPUT:url
                               queryParams:@{
                                             @"apikey": [LoginRadiusSDK apiKey],
                                             @"access_token": accessToken,
                                             @"objectname": objectname,
                                             @"updatetype":updatetype
                                             }
                                      body:payload
                         completionHandler:completion];
}


- (void)deleteCustomObjectWithObjectRecordId:(NSString*)objectRecordId
                       accessToken:(NSString*)accessToken
                    objectname:(NSString*)objectname
                 completionHandler:(LRAPIResponseHandler)completion{
    
    NSString* url = [NSString stringWithFormat:@"identity/v2/auth/customobject/%@",objectRecordId];
    [[LoginRadiusREST apiInstance] sendDELETE:url
                                  queryParams:@{
                                                @"apikey": [LoginRadiusSDK apiKey],
                                                @"access_token": accessToken,
                                                @"objectname": objectname
                                                }
                                         body:@{}
                            completionHandler:completion];
}

@end

File: Sources/CustomObjectAPI.swift
//
//  CustomObjectAPI.swift
//  LoginRadiusSwift SDK
//
//  Created by Megha Agarwal.
//  Copyright © 2023 LoginRadius Inc. All rights reserved.
//

import Foundation

public class CustomObjectAPI {
    
    
    public static let customInstance = CustomObjectAPI()
    
    
    public init() {}
    
    public func createCustomObject(withObjectName objectname: String,
                                   accessToken: String,
                                   payload: [String: Any],
                                   completionHandler: @escaping LRAPIResponseHandler) {
        
        let queryParams: [String: Any] = [
            "apikey": LoginRadiusSDK.apiKey(),
            "access_token": accessToken,
            "objectname": objectname
        ]
        
        LoginRadiusREST.apiInstance.sendPOST(url: "identity/v2/auth/customobject",
                                             queryParams: queryParams,
                                             body: payload,
                                             completionHandler: completionHandler)
    }
    
    public func getCustomObject(withObjectName objectname: String,
                                accessToken: String,
                                completionHandler: @escaping LRAPIResponseHandler) {
        
        let queryParams: [String: Any] = [
            "apikey": LoginRadiusSDK.apiKey(),
            "access_token": accessToken,
            "objectname": objectname
        ]
        
        LoginRadiusREST.apiInstance.sendGET(url: "identity/v2/auth/customobject",
                                            queryParams: queryParams,
                                            completionHandler: completionHandler)
    }
    
    public func getCustomObject(withObjectRecordId objectRecordId: String,
                                accessToken: String,
                                objectname: String,
                                completionHandler: @escaping LRAPIResponseHandler) {
        
        var url:String
            
               if(objectRecordId.isEmpty){
                  url = "identity/v2/auth/customobject/objectRecordId"
               }
               else{
                   url = "identity/v2/auth/customobject/\(objectRecordId)"
               }
        
        let queryParams: [String: Any] = [
            "apikey": LoginRadiusSDK.apiKey(),
            "access_token": accessToken,
            "objectname": objectname
        ]
        
        LoginRadiusREST.apiInstance.sendGET(url:url,
                                            queryParams: queryParams,
                                            completionHandler: completionHandler)
    }
    
    public func updateCustomObject(withObjectName objectname: String,
                                   accessToken: String,
                                   objectRecordId: String,
                                   updatetype: String,
                                   payload: [String: Any],
                                   completionHandler: @escaping LRAPIResponseHandler) {
        
        let url = "identity/v2/auth/customobject/\(objectRecordId)"
        
        let queryParams: [String: Any] = [
            "apikey": LoginRadiusSDK.apiKey(),
            "access_token": accessToken,
            "objectname": objectname,
            "updatetype": updatetype
        ]
        
        LoginRadiusREST.apiInstance.sendPUT(url:url,
                                            queryParams: queryParams,
                                            body: payload,
                                            completionHandler: completionHandler)
    }
    
    public func deleteCustomObject(withObjectRecordId objectRecordId: String,
                                   accessToken: String,
                                   objectname: String,
                                   completionHandler: @escaping LRAPIResponseHandler) {
        
        var url:String
            
               if(objectRecordId.isEmpty){
                  url = "identity/v2/auth/customobject/objectRecordId"
               }
               else{
                   url = "identity/v2/auth/customobject/\(objectRecordId)"
               }

        
        let queryParams: [String: Any] = [
            "apikey": LoginRadiusSDK.apiKey(),
            "access_token": accessToken,
            "objectname": objectname
        ]
        
        LoginRadiusREST.apiInstance.sendDELETE(url:url,
                                               queryParams: queryParams,
                                               body: [:],
                                               completionHandler: completionHandler)
    }
}

